centml/compiler/server_compilation.py

```diff
@@ -2,21 +2,63 @@
 import shutil
 import logging
 from typing import List
-from torch import Tensor
+import pickle
+import torch
 from torch.fx import GraphModule
-from hidet.runtime.compiled_graph import save_compiled_graph
 from hidet.graph.frontend import from_torch
 from hidet.graph.frontend.torch.interpreter import Interpreter
 from hidet.graph.frontend.torch.dynamo_backends import get_flow_graph, get_compiled_graph, preprocess_inputs
+from centml.compiler.hidet_compiled_model import HidetCompiledModel
 from centml.compiler.config import config_instance
 
-
 storage_path = os.path.join(config_instance.CACHE_PATH, "server")
 os.makedirs(storage_path, exist_ok=True)
 
 logger = logging.getLogger(__name__)
 
 
+# Calling the torch.fx.GraphModule will call RootModule.forward
+# Due to limitations on how the GraphModule is constructed, args are passed as a tuple and later unpacked
+class RootModule(torch.nn.Module):
+    def __init__(self, callable):
+        super().__init__()
+        self.leaf_module = callable
+
+    def forward(self, args):
+        return self.leaf_module(args)
+
+
+# We wrap the callable in a torch.nn.Module so that it can be a leaf module in the graph
+# Leaf modules avoid being traced by the torch.fx.Tracer and are treated like black boxes
+class ModuleWrapper(torch.nn.Module):
+    def __init__(self, callable):
+        super().__init__()
+        self.callable = callable
+
+    def forward(self, args):
+        return self.callable(*args)
+
+
+# Note: pickling a GraphModule won't save its Graph.
+# Instead, it stores the Tracer and re-traces to recreate the Graph when deserializing.
+class CustomTracer(torch.fx.Tracer):
+    # Don't trace the ModuleWrapper
+    def is_leaf_module(self, m, module_qualified_name):
+        if isinstance(m, ModuleWrapper):
+            return True
+        return super().is_leaf_module(m, module_qualified_name)
+
+
+# Create a torch.fx.GraphModule that wraps around `callable`.
+# `callable` is a class whose forward function gets called when we call the GraphModule's forward function
+def get_graph_module(callable):
+    module = ModuleWrapper(callable)
+    root = RootModule(module)
```

Comment:
Is it possible to make RootModule an extension of GraphModule. And move this wrapping logic to inside RootModule.__init__() ?

Reply:
I don't think so. Basically GraphModules consist a graph and a root. 
The graph is a series of torch Nodes which taken together form a python function (torch.fx optimizes this "function" this way). 
The root is a nn.Module can be thought of as the base/root of the graph.

Here, RootModule is the GraphModule's root. So the reason we can't put this logic inside an extension of GraphModule's `__init__` is because that GraphModule would be it's own root. Based on some testing, this doesn't seem possible

Reply:
we decided to not use GraphModule